File: ats-service/app/analysis/keywords.py
"""Multi-level Keyword Matching Engine."""
from typing import Dict, Any, List, Set, Tuple
from ..models.schemas import MatchedSkill, MissingSkill
from .skills import SkillTaxonomy

class KeywordMatcher:
    """Matches job requirements against resume skills across multiple precision tiers."""

    def __init__(self):
        self.taxonomy = SkillTaxonomy.get_instance()

    def match(
        self,
        resume_skills: Dict[str, Dict[str, Any]],
        jd_analysis: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Performs multi-level matching between resume skills and job requirements."""
        if not jd_analysis.get("has_jd", False):
            # If no JD is provided, treat detected resume skills as matched inventory
            matched: List[MatchedSkill] = []
            for canon, data in resume_skills.items():
                matched.append(MatchedSkill(
                    skill=data["found_term"].capitalize(),
                    match_type="exact",
                    canonical_name=canon,
                    category=data["category"]
                ))
            return {
                "matched_skills": matched,
                "missing_skills": [],
                "related_skills": [],
                "coverage_ratio": 1.0,
                "matched_count": len(matched),
                "total_target_count": len(matched),
            }

        required_targets = jd_analysis.get("required_skills", [])
        preferred_targets = jd_analysis.get("preferred_skills", [])
        all_targets = list(set(required_targets + preferred_targets))

        matched_skills: List[MatchedSkill] = []
        missing_skills: List[MissingSkill] = []
        related_skills: List[str] = []

        resume_canonicals = set(resume_skills.keys())

        # Check each target skill in the job description
        for target in all_targets:
            target_canonical = self.taxonomy.get_canonical(target) or target.lower()
            target_category = self.taxonomy.get_category(target_canonical)
            importance = "required" if target in required_targets else "preferred"

            # 1. Exact canonical or direct match
            if target_canonical in resume_canonicals:
                data = resume_skills[target_canonical]
                found_term = data["found_term"]

                # Distinguish exact text vs synonym match
                if found_term.lower() == target.lower():
                    m_type = "exact"
                elif found_term.lower().replace(".", "").replace(" ", "") == target.lower().replace(".", "").replace(" ", ""):
                    m_type = "normalized"
                else:
                    m_type = "synonym"

                matched_skills.append(MatchedSkill(
                    skill=found_term.capitalize(),
                    match_type=m_type,
                    canonical_name=target_canonical,
                    category=target_category
                ))
            else:
                # Skill is missing from resume
                missing_skills.append(MissingSkill(
                    skill=target.capitalize(),
                    importance=importance,
                    category=target_category
                ))

                # Check if candidate has a related technology in the same category
                # E.g. job wants React, candidate has Vue
                for r_canon in resume_canonicals:
                    if self.taxonomy.are_related(target_canonical, r_canon):
                        rel_info = f"{resume_skills[r_canon]['found_term'].capitalize()} (related {target_category} skill for {target.capitalize()})"
                        if rel_info not in related_skills:
                            related_skills.append(rel_info)

        total_targets = len(all_targets) if all_targets else 1
        matched_count = len(matched_skills)
        coverage_ratio = matched_count / total_targets

        return {
            "matched_skills": matched_skills,
            "missing_skills": missing_skills,
            "related_skills": related_skills[:6],
            "coverage_ratio": round(coverage_ratio, 2),
            "matched_count": matched_count,
            "total_target_count": total_targets,
        }

File: src/lib/ats/keywords.js
import {
  ALIAS_TO_CANONICAL,
  SKILL_TO_CATEGORY,
  getCanonical,
  getCategory,
  areRelated
} from './taxonomy.js';

/**
 * Deterministic Multi-tier Keyword and Skill Extraction Engine.
 */

// Pre-sorted list of terms by length descending to prioritize multi-word matches
const ALL_TERMS = Array.from(
  new Set([...Object.keys(ALIAS_TO_CANONICAL), ...Object.keys(SKILL_TO_CATEGORY)])
).sort((a, b) => b.length - a.length);

function escapeRegex(string) {
  return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

export function extractSkillsFromText(text) {
  if (!text) return {};
  const textLower = text.toLowerCase();
  const detected = {};

  for (const term of ALL_TERMS) {
    const escaped = escapeRegex(term);
    let regex;

    // Special characters handling: c++, c#, .net
    if (['c++', 'c#', '.net'].includes(term)) {
      regex = new RegExp(`(?:^|\\s|[(\\[])${escaped}(?:$|\\s|[,;.:)\\]])`, 'gi');
    } else {
      regex = new RegExp(`\\b${escaped}\\b`, 'gi');
    }

    const matches = textLower.match(regex);
    if (matches && matches.length > 0) {
      const canonical = getCanonical(term) || term;
      const category = getCategory(canonical);

      if (!detected[canonical]) {
        detected[canonical] = {
          found_term: term,
          canonical,
          category,
          count: matches.length
        };
      } else {
        detected[canonical].count += matches.length;
      }
    }
  }

  return detected;
}

export function parseJobDescription(jdText) {
  if (!jdText || !jdText.trim()) {
    return {
      has_jd: false,
      required_skills: [],
      preferred_skills: [],
      responsibilities: [],
      experience_requirement: null,
      education_requirement: null
    };
  }

  const lines = jdText.split(/\r?\n/).map((l) => l.trim()).filter(Boolean);
  const requiredSet = new Set();
  const preferredSet = new Set();
  const responsibilities = [];

  let isPreferredMode = false;
  let isRequiredMode = false;
  let isRespMode = false;

  for (const line of lines) {
    const lineLower = line.toLowerCase();

    if (/nice to have|preferred|bonus|good to have|plus/i.test(lineLower)) {
      isPreferredMode = true;
      isRequiredMode = false;
      isRespMode = false;
      continue;
    } else if (/required|requirements|must have|qualifications|what you need/i.test(lineLower)) {
      isRequiredMode = true;
      isPreferredMode = false;
      isRespMode = false;
      continue;
    } else if (/responsibilities|what you will do|role overview|duties/i.test(lineLower)) {
      isRespMode = true;
      isRequiredMode = false;
      isPreferredMode = false;
      continue;
    }

    const extracted = extractSkillsFromText(line);
    for (const canon of Object.keys(extracted)) {
      if (isPreferredMode) {
        preferredSet.add(canon);
      } else {
        requiredSet.add(canon);
      }
    }

    if (isRespMode && line.length > 20) {
      responsibilities.push(line.replace(/^[•\-* \t]+/, ''));
    }
  }

  // Fallback: if no section headers separated required vs preferred, treat all extracted skills as required
  if (requiredSet.size === 0 && preferredSet.size === 0) {
    const allFound = extractSkillsFromText(jdText);
    for (const c of Object.keys(allFound)) {
      requiredSet.add(c);
    }
  }

  const expMatch = jdText.match(/(\d+[\+]?(?:\s*-\s*\d+)?\s*(?:years|yrs)\s+(?:of\s+)?experience)/i);
  const experience_requirement = expMatch ? expMatch[1] : null;

  let education_requirement = null;
  const eduMatch = jdText.match(/([^.\n]*(?:bachelor|master|degree|computer science)[^.\n]*)/i);
  if (eduMatch) {
    education_requirement = eduMatch[1].trim();
  }

  return {
    has_jd: true,
    required_skills: Array.from(requiredSet),
    preferred_skills: Array.from(preferredSet),
    responsibilities: responsibilities.slice(0, 8),
    experience_requirement,
    education_requirement,
    raw_text: jdText
  };
}

export function matchKeywords(resumeSkills, jdAnalysis) {
  if (!jdAnalysis || !jdAnalysis.has_jd) {
    const matched = [];
    for (const [canon, data] of Object.entries(resumeSkills)) {
      matched.push({
        skill: data.found_term.charAt(0).toUpperCase() + data.found_term.slice(1),
        match_type: 'exact',
        canonical_name: canon,
        category: data.category
      });
    }
    return {
      matched_skills: matched,
      missing_skills: [],
      related_skills: [],
      coverage_ratio: 1.0,
      matched_count: matched.length,
      total_target_count: matched.length
    };
  }

  const requiredTargets = jdAnalysis.required_skills || [];
  const preferredTargets = jdAnalysis.preferred_skills || [];
  const allTargets = Array.from(new Set([...requiredTargets, ...preferredTargets]));

  const matchedSkills = [];
  const missingSkills = [];
  const relatedSkills = [];

  const resumeCanonicals = new Set(Object.keys(resumeSkills));

  for (const target of allTargets) {
    const targetCanonical = getCanonical(target) || target.toLowerCase();
    const targetCategory = getCategory(targetCanonical);
    const importance = requiredTargets.includes(target) ? 'required' : 'preferred';

    if (resumeCanonicals.has(targetCanonical)) {
      const data = resumeSkills[targetCanonical];
      const foundTerm = data.found_term;

      let matchType = 'exact';
      if (foundTerm.toLowerCase() === target.toLowerCase()) {
        matchType = 'exact';
      } else if (
        foundTerm.toLowerCase().replace(/[.\s]/g, '') === target.toLowerCase().replace(/[.\s]/g, '')
      ) {
        matchType = 'normalized';
      } else {
        matchType = 'synonym';
      }

      matchedSkills.push({
        skill: foundTerm.charAt(0).toUpperCase() + foundTerm.slice(1),
        match_type: matchType,
        canonical_name: targetCanonical,
        category: targetCategory
      });
    } else {
      missingSkills.push({
        skill: target.charAt(0).toUpperCase() + target.slice(1),
        importance,
        category: targetCategory
      });

      // Find related technologies in resume within same category
      for (const rCanon of resumeCanonicals) {
        if (areRelated(targetCanonical, rCanon)) {
          const relTerm = resumeSkills[rCanon].found_term;
          const capitalizedRel = relTerm.charAt(0).toUpperCase() + relTerm.slice(1);
          const capitalizedTarget = target.charAt(0).toUpperCase() + target.slice(1);
          const relInfo = `${capitalizedRel} (related ${targetCategory} skill for ${capitalizedTarget})`;
          if (!relatedSkills.includes(relInfo)) {
            relatedSkills.push(relInfo);
          }
        }
      }
    }
  }

  const totalTargets = Math.max(1, allTargets.length);
  const matchedCount = matchedSkills.length;
  const coverageRatio = Math.min(1.0, matchedCount / totalTargets);

  return {
    matched_skills: matchedSkills,
    missing_skills: missingSkills,
    related_skills: relatedSkills.slice(0, 6),
    coverage_ratio: Number(coverageRatio.toFixed(2)),
    matched_count: matchedCount,
    total_target_count: totalTargets
  };
}
